### PR TITLE
Fix train.py for saving sample npy

### PR DESCRIPTION
**Pull request by seldauyanik-maxim**
The validation function is fixed for the case when --save-sample option was selected in train.py.

### CODE REVIEW

**Comment by seldauyanik-maxim** [comment]
PR for solving issue: #199

**Review by ermanok — DISMISSED** [review]
Looks good.

**Review by Jake-Carter — CHANGES_REQUESTED** [review]
@seldauyanik-maxim Thanks, this fixes `--save-sample` breaking mAP **but the saved sample is all `0`'s and not usable.**

I think my original PR never got merged because we never found why we were doing it the old way.  I think `--save-sample` was intended only to be used with `--evaluate`, _not_ training.

**Findings:**
* `--save-sample` only works with `--evaluate`
* `--save-sample` + `--evaluate` still does _not_ work for object detection models

**Proposed Changes:**
* Fix `--evaluate` + `--save-sample` for object detection models
* Add evaluation script for tinyssd
* Make `--save-sample` work during training, or make it exclusive to `--evaluate` and document
* Make `--save-sample` easier to use on its own.  Ex: 
    ```bash
    python train.py --model ... --device ... --dataset SVHN --save-sample 10
    ```

**Logs:**

**Before**
Broken mAP with `--save-sample`:
```bash
(ai8x-training) jakecarter@msi2:~/repos/ai8x-training$ bash scripts/train_svhn_tinierssd.sh --deterministic --save-sample 10
Configuring device: MAX78000, simulate=False.
Log file for this run: /home/jakecarter/repos/ai8x-training/logs/2022.12.13-132923/2022.12.13-132923.log
{'start_epoch': 25, 'weight_bits': 8, 'shift_quantile': 0.995}
{'multi_box_loss': {'alpha': 2, 'neg_pos_ratio': 3}, 'nms': {'min_score': 0.2, 'max_overlap': 0.3, 'top_k': 20}}
Optimizer Type: <class 'torch.optim.adam.Adam'>
Optimizer Args: {'lr': 0.001, 'betas': (0.9, 0.999), 'eps': 1e-08, 'weight_decay': 0.0005, 'amsgrad': False}
Train dataset length: 28548

Test dataset length: 12251

Dataset sizes:
        training=28548
        validation=12251
        test=12251


Training epoch: 28548 samples (16 per mini-batch)
Epoch: [0][  200/ 1785]    Overall Loss 10.771582    Objective Loss 10.771582                                        LR 0.001000    Time 0.040218    
Epoch: [0][  400/ 1785]    Overall Loss 10.499871    Objective Loss 10.499871                                        LR 0.001000    Time 0.038819    
Epoch: [0][  600/ 1785]    Overall Loss 10.236256    Objective Loss 10.236256                                        LR 0.001000    Time 0.038270    
Epoch: [0][  800/ 1785]    Overall Loss 9.979996    Objective Loss 9.979996                                        LR 0.001000    Time 0.038077    
Epoch: [0][ 1000/ 1785]    Overall Loss 9.747530    Objective Loss 9.747530                                        LR 0.001000    Time 0.038203    
Epoch: [0][ 1200/ 1785]    Overall Loss 9.536320    Objective Loss 9.536320                                        LR 0.001000    Time 0.038256    
Epoch: [0][ 1400/ 1785]    Overall Loss 9.344865    Objective Loss 9.344865                                        LR 0.001000    Time 0.038291    
Epoch: [0][ 1600/ 1785]    Overall Loss 9.170985    Objective Loss 9.170985                                        LR 0.001000    Time 0.038316    
Epoch: [0][ 1785/ 1785]    Overall Loss 9.019536    Objective Loss 9.019536                                        LR 0.001000    Time 0.038326    
--- validate (epoch=0)-----------
12251 samples (16 per mini-batch)
{'multi_box_loss': {'alpha': 2, 'neg_pos_ratio': 3}, 'nms': {'min_score': 0.2, 'max_overlap': 0.3, 'top_k': 20}}
==> Saving sample at index 10 to sample_svhn_74.npy
==> Best [mAP: 0.000000   vloss: 0.000000   Sparsity:0.00   Params: 335520 on epoch: 0]
```

Saved sample is all `0`'s:
```python
>>> import numpy as np
>>> np.load("sample_svhn_74.npy")
array([[[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]]])
```
---

**After**
Fixed mAP with `--save-sample`:
```bash
(ai8x-training) jakecarter@msi2:~/repos/ai8x-training$ bash scripts/train_svhn_tinierssd.sh --deterministic --save-sample 10
Configuring device: MAX78000, simulate=False.
Log file for this run: /home/jakecarter/repos/ai8x-training/logs/2022.12.13-132322/2022.12.13-132322.log
{'start_epoch': 25, 'weight_bits': 8, 'shift_quantile': 0.995}
{'multi_box_loss': {'alpha': 2, 'neg_pos_ratio': 3}, 'nms': {'min_score': 0.2, 'max_overlap': 0.3, 'top_k': 20}}
Optimizer Type: <class 'torch.optim.adam.Adam'>
Optimizer Args: {'lr': 0.001, 'betas': (0.9, 0.999), 'eps': 1e-08, 'weight_decay': 0.0005, 'amsgrad': False}
Train dataset length: 28548

Test dataset length: 12251

Dataset sizes:
        training=28548
        validation=12251
        test=12251


Training epoch: 28548 samples (16 per mini-batch)
Epoch: [0][  200/ 1785]    Overall Loss 10.771582    Objective Loss 10.771582                                        LR 0.001000    Time 0.039856    
Epoch: [0][  400/ 1785]    Overall Loss 10.499871    Objective Loss 10.499871                                        LR 0.001000    Time 0.038550    
Epoch: [0][  600/ 1785]    Overall Loss 10.236256    Objective Loss 10.236256                                        LR 0.001000    Time 0.038026    
Epoch: [0][  800/ 1785]    Overall Loss 9.979996    Objective Loss 9.979996                                        LR 0.001000    Time 0.037745    
Epoch: [0][ 1000/ 1785]    Overall Loss 9.747530    Objective Loss 9.747530                                        LR 0.001000    Time 0.037652    
Epoch: [0][ 1200/ 1785]    Overall Loss 9.536320    Objective Loss 9.536320                                        LR 0.001000    Time 0.037678    
Epoch: [0][ 1400/ 1785]    Overall Loss 9.344865    Objective Loss 9.344865                                        LR 0.001000    Time 0.037685    
Epoch: [0][ 1600/ 1785]    Overall Loss 9.170985    Objective Loss 9.170985                                        LR 0.001000    Time 0.037685    
Epoch: [0][ 1785/ 1785]    Overall Loss 9.019536    Objective Loss 9.019536                                        LR 0.001000    Time 0.037666    
--- validate (epoch=0)-----------
12251 samples (16 per mini-batch)
{'multi_box_loss': {'alpha': 2, 'neg_pos_ratio': 3}, 'nms': {'min_score': 0.2, 'max_overlap': 0.3, 'top_k': 20}}
==> Saving sample at index 10 to sample_svhn_74.npy
Epoch: [0][  200/  766]    Loss 7.648111    mAP 0.680983    
Epoch: [0][  400/  766]    Loss 7.646800    mAP 0.677540    
Epoch: [0][  600/  766]    Loss 7.649039    mAP 0.675232    
Epoch: [0][  766/  766]    Loss 7.648984    mAP 0.672360    
==> mAP: 0.67236    Loss: 7.649

==> Best [mAP: 0.672360   vloss: 7.648984   Sparsity:0.00   Params: 335520 on epoch: 0]
```
Saved sample is still all `0`'s
```python
>>> import numpy as np
>>> np.load("sample_svhn_74.npy")
array([[[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
```
---
**Using `--save-sample` + `--evaluate`**  (Used on tinierssd checkpoint after training 1 epoch)
mAP is fixed:
```bash
(ai8x-training) jakecarter@msi2:~/repos/ai8x-training$ python train.py --deterministic --model ai85tinierssd --use-bias --dataset SVHN_74 --device MAX78000 --obj-detection --obj-detection-params parameters/obj_detection_params_svhn.yaml --qat-policy policies/qat_policy_svhn.yaml --exp-load-weights-from logs/2022.12.13-132923/checkpoint.pth.tar --evaluate --save-sample 10
Configuring device: MAX78000, simulate=False.
Log file for this run: /home/jakecarter/repos/ai8x-training/logs/2022.12.13-133638/2022.12.13-133638.log
{'start_epoch': 25, 'weight_bits': 8, 'shift_quantile': 0.995}
{'multi_box_loss': {'alpha': 2, 'neg_pos_ratio': 3}, 'nms': {'min_score': 0.2, 'max_overlap': 0.3, 'top_k': 20}}
=> loading checkpoint logs/2022.12.13-132923/checkpoint.pth.tar
=> Checkpoint contents:
+----------------------+-------------+---------------+
| Key                  | Type        | Value         |
|----------------------+-------------+---------------|
| arch                 | str         | ai85tinierssd |
| compression_sched    | dict        |               |
| epoch                | int         | 0             |
| extras               | dict        |               |
| optimizer_state_dict | dict        |               |
| optimizer_type       | type        | Adam          |
| state_dict           | OrderedDict |               |
+----------------------+-------------+---------------+

=> Checkpoint['extras'] contents:
+--------------+--------+---------+
| Key          | Type   |   Value |
|--------------+--------+---------|
| best_epoch   | int    |       0 |
| best_mAP     | float  |       0 |
| best_top1    | float  |       0 |
| current_mAP  | float  |       0 |
| current_top1 | float  |       0 |
+--------------+--------+---------+

Loaded compression schedule from checkpoint (epoch 0)
=> loaded 'state_dict' from checkpoint 'logs/2022.12.13-132923/checkpoint.pth.tar'
Optimizer Type: <class 'torch.optim.sgd.SGD'>
Optimizer Args: {'lr': 0.1, 'momentum': 0.9, 'dampening': 0, 'weight_decay': 0.0001, 'nesterov': False}
Train dataset length: 28548

Test dataset length: 12251

Dataset sizes:
        training=25694
        validation=2854
        test=12251
--- test ---------------------
12251 samples (256 per mini-batch)
{'multi_box_loss': {'alpha': 2, 'neg_pos_ratio': 3}, 'nms': {'min_score': 0.2, 'max_overlap': 0.3, 'top_k': 20}}
==> Saving sample at index 10 to sample_svhn_74.npy
Test: [   10/   48]    Loss 7.662502    mAP 0.670840
Test: [   20/   48]    Loss 7.657196    mAP 0.667955
Test: [   30/   48]    Loss 7.651973    mAP 0.669851
Test: [   40/   48]    Loss 7.648975    mAP 0.668783
Test: [   48/   48]    Loss 7.648786    mAP 0.667676
==> mAP: 0.66768    Loss: 7.649
```
Saved sample is still all `0`'s
```python
>>> import numpy as np
>>> np.load("sample_svhn_74.npy")
array([[[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]]])
```
---
Using `--save-sample` during training for **catsdogs**:
```bash
(ai8x-training) jakecarter@msi2:~/repos/ai8x-training$ bash scripts/train_catsdogs.sh --save-sample 10
Configuring device: MAX78000, simulate=False.
Log file for this run: /home/jakecarter/repos/ai8x-training/logs/2022.12.13-134901/2022.12.13-134901.log
{'start_epoch': 10, 'weight_bits': 8}
Optimizer Type: <class 'torch.optim.adam.Adam'>
Optimizer Args: {'lr': 0.001, 'betas': (0.9, 0.999), 'eps': 1e-08, 'weight_decay': 0.0, 'amsgrad': False}
augmented folder exits. Remove if you want to regenerate
Dataset sizes:
        training=72000
        validation=8000
        test=5000
Reading compression schedule from: policies/schedule-catsdogs.yaml


Training epoch: 72000 samples (256 per mini-batch)
Epoch: [0][   10/  282]    Overall Loss 0.691738    Objective Loss 0.691738                                        LR 0.001000    Time 0.146905    
Epoch: [0][   20/  282]    Overall Loss 0.684985    Objective Loss 0.684985                                        LR 0.001000    Time 0.104445    
Epoch: [0][   30/  282]    Overall Loss 0.680830    Objective Loss 0.680830                                        LR 0.001000    Time 0.090167    
Epoch: [0][   40/  282]    Overall Loss 0.674306    Objective Loss 0.674306                                        LR 0.001000    Time 0.083142    
Epoch: [0][   50/  282]    Overall Loss 0.670331    Objective Loss 0.670331                                        LR 0.001000    Time 0.078826    
Epoch: [0][   60/  282]    Overall Loss 0.665347    Objective Loss 0.665347                                        LR 0.001000    Time 0.076014    
Epoch: [0][   70/  282]    Overall Loss 0.660816    Objective Loss 0.660816                                        LR 0.001000    Time 0.074051    
Epoch: [0][   80/  282]    Overall Loss 0.655555    Objective Loss 0.655555                                        LR 0.001000    Time 0.072563    
Epoch: [0][   90/  282]    Overall Loss 0.651566    Objective Loss 0.651566                                        LR 0.001000    Time 0.071385    
Epoch: [0][  100/  282]    Overall Loss 0.646706    Objective Loss 0.646706                                        LR 0.001000    Time 0.070490    
Epoch: [0][  110/  282]    Overall Loss 0.640652    Objective Loss 0.640652                                        LR 0.001000    Time 0.069717    
Epoch: [0][  120/  282]    Overall Loss 0.636553    Objective Loss 0.636553                                        LR 0.001000    Time 0.069050    
Epoch: [0][  130/  282]    Overall Loss 0.631721    Objective Loss 0.631721                                        LR 0.001000    Time 0.068531    
Epoch: [0][  140/  282]    Overall Loss 0.627428    Objective Loss 0.627428                                        LR 0.001000    Time 0.068067    
Epoch: [0][  150/  282]    Overall Loss 0.623933    Objective Loss 0.623933                                        LR 0.001000    Time 0.067647    
Epoch: [0][  160/  282]    Overall Loss 0.620590    Objective Loss 0.620590                                        LR 0.001000    Time 0.067287    
Epoch: [0][  170/  282]    Overall Loss 0.617645    Objective Loss 0.617645                                        LR 0.001000    Time 0.067030    
Epoch: [0][  180/  282]    Overall Loss 0.614715    Objective Loss 0.614715                                        LR 0.001000    Time 0.066758    
Epoch: [0][  190/  282]    Overall Loss 0.610853    Objective Loss 0.610853                                        LR 0.001000    Time 0.066503    
Epoch: [0][  200/  282]    Overall Loss 0.608342    Objective Loss 0.608342                                        LR 0.001000    Time 0.066297    
Epoch: [0][  210/  282]    Overall Loss 0.606145    Objective Loss 0.606145                                        LR 0.001000    Time 0.066116    
Epoch: [0][  220/  282]    Overall Loss 0.603616    Objective Loss 0.603616                                        LR 0.001000    Time 0.065935    
Epoch: [0][  230/  282]    Overall Loss 0.600507    Objective Loss 0.600507                                        LR 0.001000    Time 0.065772    
Epoch: [0][  240/  282]    Overall Loss 0.597720    Objective Loss 0.597720                                        LR 0.001000    Time 0.065679    
Epoch: [0][  250/  282]    Overall Loss 0.595096    Objective Loss 0.595096                                        LR 0.001000    Time 0.065563    
Epoch: [0][  260/  282]    Overall Loss 0.592405    Objective Loss 0.592405                                        LR 0.001000    Time 0.065452    
Epoch: [0][  270/  282]    Overall Loss 0.589936    Objective Loss 0.589936                                        LR 0.001000    Time 0.065336    
Epoch: [0][  280/  282]    Overall Loss 0.587741    Objective Loss 0.587741                                        LR 0.001000    Time 0.065239    
Epoch: [0][  282/  282]    Overall Loss 0.587261    Objective Loss 0.587261    Top1 73.437500    LR 0.001000    Time 0.065157    
--- validate (epoch=0)-----------
8000 samples (256 per mini-batch)
==> Saving sample at index 10 to sample_cats_vs_dogs.npy
Epoch: [0][   10/   32]    Loss 0.535097    Top1 73.242188    
Epoch: [0][   20/   32]    Loss 0.526639    Top1 73.613281    
Epoch: [0][   30/   32]    Loss 0.535637    Top1 72.916667    
Epoch: [0][   32/   32]    Loss 0.532912    Top1 72.937500    
==> Top1: 72.938    Loss: 0.533

==> Confusion:
[[2540 1341]
 [ 824 3295]]

==> Best [Top1: 72.938   Sparsity:0.00   Params: 57776 on epoch: 0]
```
**(saved sample is broken)**
```python
>>> import numpy as np
>>> np.load("sample_cats_vs_dogs.npy")
array([[[ 0,  0,  0, ...,  0,  0,  0],
        [ 0,  0,  0, ...,  0,  0,  0],
        [ 0,  0,  0, ...,  0,  0,  0],
        ...,
        [-1, -1, -1, ...,  0,  0,  0],
        [-1, -1, -1, ...,  0,  0,  0],
        [-1, -1, -1, ...,  0,  0,  0]],

       [[ 0,  0,  0, ...,  0,  0,  0],
        [ 0,  0,  0, ...,  0,  0,  0],
        [ 0,  0,  0, ...,  0,  0,  0],
        ...,
        [-1, -1, -1, ...,  0,  0,  0],
        [-1, -1, -1, ...,  0,  0,  0],
        [-1, -1, -1, ...,  0,  0,  0]],

       [[ 0,  0,  0, ...,  0,  0,  0],
        [ 0,  0,  0, ...,  0,  0,  0],
        [ 0,  0,  0, ...,  0,  0,  0],
        ...,
        [-1, -1, -1, ...,  0,  0,  0],
        [-1, -1, -1, ...,  0,  0,  0],
        [-1, -1, -1, ...,  0,  0,  0]]])
```
Using `--save-sample` + `--evaluate` on cats_vs_dogs:
```bash
(ai8x-training) jakecarter@msi2:~/repos/ai8x-training$ bash scripts/evaluate_catsdogs.sh --save-sample 10
Configuring device: MAX78000, simulate=True.
Log file for this run: /home/jakecarter/repos/ai8x-training/logs/2022.12.13-135211/2022.12.13-135211.log
{'start_epoch': 10, 'weight_bits': 8}
=> loading checkpoint ../ai8x-synthesis/trained/ai85-catsdogs-qat8-q.pth.tar
=> Checkpoint contents:
+----------------------+-------------+-----------+
| Key                  | Type        | Value     |
|----------------------+-------------+-----------|
| arch                 | str         | ai85cdnet |
| compression_sched    | dict        |           |
| epoch                | int         | 148       |
| extras               | dict        |           |
| optimizer_state_dict | dict        |           |
| optimizer_type       | type        | Adam      |
| state_dict           | OrderedDict |           |
+----------------------+-------------+-----------+

=> Checkpoint['extras'] contents:
+-----------------+--------+------------------+
| Key             | Type   | Value            |
|-----------------+--------+------------------|
| best_epoch      | int    | 148              |
| best_top1       | float  | 93.2282336080258 |
| clipping_method | str    | MAX_BIT_SHIFT    |
| current_top1    | float  | 93.2282336080258 |
+-----------------+--------+------------------+

Loaded compression schedule from checkpoint (epoch 148)
=> loaded 'state_dict' from checkpoint '../ai8x-synthesis/trained/ai85-catsdogs-qat8-q.pth.tar'
Optimizer Type: <class 'torch.optim.sgd.SGD'>
Optimizer Args: {'lr': 0.1, 'momentum': 0.9, 'dampening': 0, 'weight_decay': 0.0001, 'nesterov': False}
augmented folder exits. Remove if you want to regenerate
Dataset sizes:
        training=72000
        validation=8000
        test=5000
--- test ---------------------
5000 samples (256 per mini-batch)
==> Saving sample at index 10 to sample_cats_vs_dogs.npy
Test: [   10/   20]    Loss 0.085216    Top1 97.929688    
Test: [   20/   20]    Loss 0.085654    Top1 98.000000    
==> Top1: 98.000    Loss: 0.086

==> Confusion:
[[2437   63]
 [  37 2463]]
```
Saved sample is valid
```python
>>> import numpy as np
>>> np.load("sample_cats_vs_dogs.npy")
array([[[ -94,  -64,  -39, ...,   61,   48,   70],
        [ -89,  -87,  -37, ...,   45,   25,   68],
        [ -82,  -59,  -33, ...,   30,   27,   41],
        ...,
        [ -98, -100, -105, ...,  -20,    1,  -11],
        [-109,  -91,  -87, ...,  -11,  -36,   -5],
        [ -99, -100,  -91, ...,  -26,  -74,  -49]],

       [[ -79,  -48,  -20, ...,   53,   51,   75],
        [ -68,  -65,  -14, ...,   46,   26,   64],
        [ -59,  -36,   -9, ...,   45,   38,   43],
        ...,
        [ -81,  -81,  -84, ...,  -14,    9,   -1],
        [ -96,  -68,  -51, ...,    7,  -22,   11],
        [ -85,  -75,  -52, ...,   16,  -37,  -11]],

       [[-109,  -91,  -83, ...,  -17,  -22,    3],
        [-100, -107,  -75, ...,  -29,  -49,  -10],
        [ -98,  -85,  -71, ...,  -33,  -40,  -33],
        ...,
        [-114, -112, -110, ...,  -79,  -58,  -71],
        [-110,  -98, -100, ...,  -62,  -83,  -67],
        [-104, -109, -106, ...,  -59, -101,  -96]]])
```

File: train.py [hunk 1]
```diff
@@ -194,6 +194,9 @@ def main():
             print('WARNING: Initial learning rate (--lr) not set, selecting 0.1.')
         args.lr = 0.1
 
+    if args.generate_sample is not None and not args.act_mode_8bit:
+            print('WARNING: Cannot save sample in training mode, ignoring --save-sample option.')
```
Comment (rotx-maxim) [inline train.py:198]: Please check the spaces (should be 4 spaces)

Comment (seldauyanik-maxim) [reply train.py]: Done.

File: train.py [hunk 2]
```diff
@@ -194,6 +194,9 @@ def main():
             print('WARNING: Initial learning rate (--lr) not set, selecting 0.1.')
         args.lr = 0.1
 
+    if args.generate_sample is not None and not args.act_mode_8bit:
+        print('WARNING: Cannot save sample in training mode, ignoring --save-sample option.')
```
Comment (Jake-Carter) [inline train.py:198]: Can we use `'WARNING: Cannot save sample in training mode, ignoring --save-sample option.  Use with --evaluate instead.'` ?

**Review by Jake-Carter — CHANGES_REQUESTED** [review]
Thanks Selda, works great.  One very minor suggestion below before merge

File: train.py (hunk 2 above)
Comment (seldauyanik-maxim) [reply train.py]: Done in the last commit

**Review by Jake-Carter — APPROVED** [review]

**Pull request merged**